add support for numpy 1.18.1

munkm opened the pull request:
## PR Summary

This should fix the issues reported in yt-project/yt#2445 caused by updates in our upstream packages. 

For the yt_units errors I followed the same approach taken in yt-project/unyt#115. Thanks @ngoldbaum; your work made this a lot easer for me. 

## PR Checklist

- [x] Code passes flake8 checker

ngoldbaum commented:
Thank you for backporting that!

Maybe we should bump the numpy version in `test_requirements.txt` too?

munkm commented:
You're welcome! 🎉 It was a fun learning experience. 

So we don't set the version for numpy in test_requirements, but I've bumped the version in the travis config file!

neutrinoceros commented on yt/units/tests/test_ytarray.py:
```diff
@@ -828,7 +828,10 @@ def unary_ufunc_comparison(ufunc, a):
 
 def binary_ufunc_comparison(ufunc, a, b):
-    out = a.copy()
+    if ufunc in [np.divmod]:
```
```suggestion
    if ufunc is np.divmod:
```
No strong opinion about this though. I get that your syntax scales better in case we need to add other special cases in the future.

neutrinoceros commented on yt/units/tests/test_ytarray.py:
```diff
@@ -860,7 +863,10 @@ def binary_ufunc_comparison(ufunc, a, b):
+        assert isinstance(out, tuple)
+        assert len(out) == len(ret)
+        for o, r in zip(out, ret):
+            assert_array_equal(r, o)
```
```suggestion
            assert_array_equal(o, r)
```
just a matter of consistency in ordering.

neutrinoceros commented on yt/units/yt_array.py:
```diff
@@ -1370,9 +1372,25 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+                    out = tuple(outs)
+                else:
+                    out_element = out_orig[0]
+                    if out_element.dtype.kind in ("u", "i"):
```
```suggestion
                    if out_element.dtype.kind in "ui":
```
is equivalent but is arguably harder to read.

ngoldbaum commented:
Hey Clement,

All of the changes to yt.units are backported from unyt and probably shouldn’t be changed to make future backports (if any) easier.

ngoldbaum commented on yt/visualization/color_maps.py:
```diff
@@ -548,10 +548,10 @@ def make_colormap(ctuple_list, name=None, interpolate=True):
         # Interpolate the R, G, and B channels from one color to the next
         # Use np.round to make sure you're on a discrete index
-        interval = np.round(next_index)-np.round(rolling_index)
+        interval = int(np.round(next_index)-np.round(rolling_index))
```
I disagree that this is any clearer.

neutrinoceros commented:
Oh right, my mistake. I did not realise the nature of the backport. Sorry !

munkm commented:
Ok, I've done the following:
- removed testing on travis for Python 2.7 with newer versions of upstream packages
- retained testing minimal versions for py2.7
- added two new test jobs in travis for minimal versions of both unit tests and answer tests in python 3.6

We do have a test job running python 3.5, which is set to be sundown in September. Do we want to remove this job and add a new one for py 3.8? (note we also don't have tests on travis for py3.7 yet either)

ngoldbaum commented:
> retained testing minimal versions for py2.7

Why keep these? I don’t think we need to have any python2.7 testing at all.

Dropping 3.5 and adding 3.7 and 3.8 makes sense to me.

That said, I think there only needs to be one minimal dependency version job, with `MINIMAL=1`, that runs on python3.6 with the earliest known supported combination of packages we can easily install on CI.

I don’t think there’s enough marginal benefit to justify the additional complication of doing the minimal tests on multiple python versions, especially now that we’ve dropped python2.7 and don’t have to worry about 2/3 compatibility anymore.

munkm commented:
> Why keep these? I don’t think we need to have any python2.7 testing at all.

I'm happy to remove them if we decide that's best. I mostly kept them because we could catch backwards incompatibilities with them. But since we're dropping py2 support it makes sense to remove them. If we do, then there will only be two minimal dependency version jobs (for unit tests and answer tests in python3) and the issue with multiple`MINIMAL` definitions will be resolved.

neutrinoceros commented:
The test_ufuncs test is now failing against python 3.6 (minimal deps run), though you clearly bumped the minimal numpy version to 1.18.1... I'm not getting it.

ngoldbaum commented:
No, the minimal numpy is 1.12 on 3.6.

@mmunk it might be easiest to only run these multiple-output ufunc tests on sufficiently new numpy. Another option would be to bump the minimum required numpy to match unyt (1.13), those tests seem to pass for unyt on numpy 1.13.
